features/filesystem/fat/FATFileSystem.cpp

```diff
@@ -660,7 +660,24 @@ void FATFileSystem::dir_seek(fs_dir_t dir, off_t offset) {
     FATFS_DIR *dh = static_cast<FATFS_DIR*>(dir);
 
     lock();
-    dh->index = offset;
+    if (offset < dh->index) {
```

Comment:
This optimisation assumes that index is monotonic increasing (which appears to be the case in the current version of ChaN FatFs).  If not, rewind unconditionally and use an equality test in the while loop.